cdap-ui/app/directives/dag-plus/my-dag-ctrl.js

```diff
@@ -74,11 +75,156 @@ angular.module(PKG.name + '.commons')
+      start: () => {
+        vm.selectedNode = [];
+      },
+      move: ({added, removed}) => {
+        const selectedNodes = $scope.nodes.filter(node => {
+          if (added.indexOf(node.name) !== -1) {
+            return true;
+          }
+          return false;
+        });
+
+        /**
+         * This has to be efficient for us to be able to handle large pipelines.
+         * 
+         * Current implementation:
+         * 
+         * I/P : nodes added , nodes removed We already have selected nodes.
+         * 1. We iterate through selected nodes and remove the nodes that are removed from selection
+         * 2. Create a selected nodes map with nodeid as key
+         * 3. Get the adjacency map for the current graph
+         * 4. Then iterate through selected nodes and for each node get all nodes connected to it from the adjacency map
+         * 5. In the iteration if both the current selected node and the nodes connected to it are
+         *    in the list of selected nodes then select the connection. This is where we use the
+         *    selectedNodesMap to make a lookup.
+         * 
+         */
+        vm.selectedNode = vm.selectedNode.filter(node => removed.indexOf(node.name) === -1).concat(selectedNodes);
+        const selectedNodesMap = {};
+        vm.selectedNode.forEach(node => selectedNodesMap[node.name] = true);
+        const adjacencyMap = DAGPlusPlusNodesStore.getAdjacencyMap();
+        clearConnectionsSelection();
+        vm.selectedNode.forEach(({name}) => {
+          const connectedNodes = adjacencyMap[name];
+          if (!Array.isArray(connectedNodes)) {
+            return;
+          }
+          const connectionsFromSource = vm.instance.getConnections({sourceId: name});
+          connectedNodes.forEach(nodeId => {
+            if (!selectedNodesMap[nodeId]) {
+              return;
+            }
+            const connObj = connectionsFromSource.find(conn => conn.targetId === nodeId);
+            if (connObj) {
+              toggleConnection(connObj);
+            }
+          });
+        });
+      },
+      end: () => {
+        // console.log('selection end');
+      }
+    };
     const repaintTimeoutsMap = {};
 
+    vm.pipelineArtifactType = HydratorPlusPlusConfigStore.getAppType();
+    vm.onPipelineContextMenuPaste = ({nodes, connections}) => {
+      if (!Array.isArray(nodes) || !Array.isArray(connections)) {
+        return;
+      }
+      let {nodes: newNodes, connections: newConnections} = sanitizeNodesAndConnectionsBeforePaste({nodes, connections});
+      newNodes = [...$scope.nodes, ...newNodes];
+      newConnections  = [...$scope.connections, ...newConnections];
+      DAGPlusPlusNodesActionsFactory.createGraphFromConfig(newNodes, newConnections);
+      vm.instance.unbind('connection');
+      vm.instance.detachEveryConnection();
+      $scope.nodes = DAGPlusPlusNodesStore.getNodes();
+      $scope.connections = DAGPlusPlusNodesStore.getConnections();
+      init();
```

Comment:
this will essentially rerender the pipeline from scratch? Can't we just add the necessary nodes and connections?

Reply:
Unfortunately I don't have a solution for this. The way we bind events with jsplumb and register endpoints warrants this. Will need to find a better way to do this